server.js

```diff
@@ -27,12 +28,132 @@ const app = express()
 app.use(cors())
 app.use(bodyParser.json())
 
-// Start defining your routes here
+// Mongoose models 
+const Artist = new mongoose.model('Artist', {
+  name: String
+});
+
+const Tracks = new mongoose.model('Tracks', {
+  id: Number,
+  trackName: String,
+  genre: String,
+  bpm: Number,
+  energy: Number,
+  length: Number,
+  popularity: Number,
+  artistName: {
+    type: mongoose.Schema.Types.ObjectId,
+    ref: 'Artist'
+  }
+});
+
+// Populating database 
+
+if (process.env.RESET_DATABASE) {
+  const populateDatabase = async () => {
+    // Prevent getting multiple same responses, this starts with deleting what is existing
+    await Tracks.deleteMany();
+    await Artist.deleteMany();
+
+    // topMusicData.forEach(item => {
+    //   const newTracks = new Tracks(item)
+    //   newTracks.save();
+    // });
+
+    // Create an array with unique tracks from topMusicData 
+    const allArtists = topMusicData.map((item) => item.artistName)
+    const uniqueArtists = Array.from(new Set(allArtists));
+
+    // Create an array of artist object following the mongoose model 
+    let artists = [];
+
+    uniqueArtists.forEach(async artist => {
+      const newArtist = new Artist({ name: artist });
+
+      artists.push(newArtist);
+      await newArtist.save();
+    });
+
+    //Creates an array of the tracks objects according to the mongoose model
+    topMusicData.forEach(async trackItem => {
+      const newTrack = new Tracks({
+        ...trackItem,
+        artistName: artists.find((artist) => artist.name === trackItem.artistName)
+      });
+      await newTrack.save();
+    })
+  };
+  populateDatabase();
+};
+
+// A middleware to handle server connections errors 
+app.use((req, res, next) => {
+    if (mongoose.connection.readyState === 1) {
+      next();
+    } else {
+      res.status(500).json({ error: "Service unavailable" });
+  }
+});
+
+// Routes 
 app.get('/', (req, res) => {
-  res.send('Hello world')
-})
+  res.send('Welcome to the music API')
+
+});
+
+// Route to all tracks
+app.get('/tracks', async (req, res) => {
+  const allTracks = await Tracks.find();
+  res.json(allTracks)
+});
+
+// Route to find a specific track per id 
+app.get('/tracks/:id', async (req, res) => {
+  try {
+  const singleTracks = await Tracks.findOne({ id: req.params.id })
+  if (singleTracks) {
+    res.json(singleTracks);
+  } else {
+    res.status(404).json({ error: 'Could not find this track' });
+  }
+} catch (error) {
+  res.status(400).json({ error: 'Invalid track id' });
+}
+});
+
+// Route to search by tracknamr
+app.get('/tracks/:name', async (req, res) => {
+  const { name } = req.params;
+  const singleName = await Tracks.findOne({ trackName: name })
+  res.json(singleName);
+});
+
+// Limiting returning data
+app.get('/tracks20', async (req, res) => {
+  const allTracks = await Tracks.find(req.query).skip(20).limit(20);
+  res.json(allTracks);
+});
```

Comment:
I just wonder why you use .skip(20), I think it is enough with .limit(20) otherwise you skip the 20 first tracks and start from track number 21. Maybe you also can add some sorting here? Like sorting on reviews and show the 20 most popular tracks or anything else? I think that makes it more useful for the user :)

Reply:
But very good that you have tried the mongoose methods :)